### PR TITLE
[lldb-dap] Support column breakpoints

### PR DESCRIPTION
**Pull request by vogelsgesang**
This commit adds support for column breakpoints to lldb-dap.

To do so, support for `breakpointLocations` was added. To find all available breakpoint positions, we iterate over the line table.

The `setBreakpoints` request already forwarded the column correctly to `SBTarget::BreakpointCreateByLocation`. However, `SourceBreakpointMap` did not keep track of multiple breakpoints in the same line. To do so, the `SourceBreakpointMap` is now indexed by line+column instead of by line only.

See http://jonasdevlieghere.com/post/lldb-column-breakpoints/ for a high-level introduction to column breakpoints.

Screen recording showing the behavior:

https://github.com/user-attachments/assets/ee5129e1-b9af-44e3-acd3-c6e9bbe65a06

### CODE REVIEW

**Comment by vogelsgesang** [comment]
Screen recording showing the behavior:

https://github.com/user-attachments/assets/ee5129e1-b9af-44e3-acd3-c6e9bbe65a06

File: llvm/include/llvm/DebugInfo/DWARF/DWARFDebugLine.h
Comment (dwblaikie) [inline llvm/include/llvm/DebugInfo/DWARF/DWARFDebugLine.h:81]: Could you commit this unrelated change separately (doesn't need precommit review/a PR, can be pushed directly if you have commit access)

File: lldb/tools/lldb-dap/lldb-dap.cpp
Comment (ashgti) [inline lldb/tools/lldb-dap/lldb-dap.cpp:1026]: Do you know how often VSCode makes this request? Is this something that should be cached once we calculate the values? I'm not sure how often this is called or how expensive it is to look up the compilation unit by path.

**Review by ashgti — APPROVED** [review]
LGTM

File: lldb/tools/lldb-dap/lldb-dap.cpp
Comment (vogelsgesang) [reply lldb/tools/lldb-dap/lldb-dap.cpp:1026]: VSCode makes this request after setting a normal line-breakpoint. In particular, it does not send this request on hovers.

As such, we should be fine as long as the performance of this call is in the same ball park as the performance of the `SetBreakpoint` call itself.

The high-level algorithm in this PR should be fine, given that the `SetBreakpoint` itself also iterates the list in a very similar way (see `LineTable::FindLineEntryIndexByFileIndexImpl`.

The implementation of that algorithm might be sub-optimal. Given that we go through the `SB*` interfaces, we are allocating a lot of `shared_ptr`s afaict. It might make sense to implement a `FindPotentialBreakPointsInRange` method directly in the `LineTable`. There we could iterate and filter the line-table without allocating a `SBLineEntry` instance for each entry in the LineTable.

I would appreciate some guidance on the 2nd point: Should I introduce a `LineTable::FindPotentialBreakPointsInRange`?

CC @JDevlieghere

**Comment by jeffreytan81** [comment]
@vogelsgesang, does this feature require compiling target with `-gcolumn-info` to work? If so, any reason I did not see `-gcolumn-info` in the testcase Makefile?

**Comment by vogelsgesang** [comment]
> does this feature require compiling target with -gcolumn-info to work? 

Apparently, `-gcolumn-info` is on by default. If I pass `-gno-column-info`, the test case actually fails. But as long as I don't pass anything, column info seems to be included in the debug info.

File: lldb/packages/Python/lldbsuite/test/tools/lldb-dap/lldbdap_testcase.py
Comment (walter-erquinigo) [inline lldb/packages/Python/lldbsuite/test/tools/lldb-dap/lldbdap_testcase.py:244]: use `self.assertTrue`, which is more idiomatic in our codebase

File: lldb/test/API/tools/lldb-dap/breakpoint/TestDAP_setBreakpoints.py
Comment (walter-erquinigo) [inline lldb/test/API/tools/lldb-dap/breakpoint/TestDAP_setBreakpoints.py:273]: +1

**Review by walter-erquinigo — APPROVED** [review]
beautiful

File: lldb/tools/lldb-dap/lldb-dap.cpp
Comment (vogelsgesang) [reply lldb/tools/lldb-dap/lldb-dap.cpp:1026]: I noticed that I will likely need a very similar, yet slightly different logic for "Go To" / "Jump" support in lldb-dap.

As such, I would defer adding a `FindPotentialBreakpointsInRange` until after also implementing "Go To", so I can hopefully unify both of them into a single, new `FindLineEntriesInSourceRange` function used for both "Go To" and column breakpoints.

My plan of action would be:
1. merge this commit as is
2. implement and merge "Go To". Also using the `SB*` APIs
3. add a `LineEntry::FindLineEntriesInSourceRange` function, plump it through the `SB*` APIs and use it from both (1) and (2)

Please let me know in case you disagree with that plan, @ashgti

Comment (ashgti) [reply lldb/tools/lldb-dap/lldb-dap.cpp:1026]: That sounds reasonable to me

File: lldb/tools/lldb-dap/lldb-dap.cpp
Comment (labath) [inline lldb/tools/lldb-dap/lldb-dap.cpp:1054]: I don't think this will be correct for header files. You also need to check that `line_entry.GetFileSpec()` matches the file you're searching for.

That said, I'm somewhat worried about the performance of this code. What's the largest file you've tested it on?

Doing the lookup in the lldb internal representation could be faster (at least we wouldn't need to do the filespec comparison each time, as we could translate it to a "support file index"). For that, I guess we'd need to expose a new SB function to support these kinds of queries.

Comment (vogelsgesang) [reply lldb/tools/lldb-dap/lldb-dap.cpp:1054]: > I don't think this will be correct for header files. You also need to check that line_entry.GetFileSpec() matches the file you're searching for.

Good point, will fix

> That said, I'm somewhat worried about the performance of this code. What's the largest file you've tested it on?

re performance considerations, also see discussion in other thread https://github.com/llvm/llvm-project/pull/113787#discussion_r1819404026

Afaict, the algorithm itself has a similar complexity to setting a breakpoint. I.e., I am turning `O(n)` into a `O(2*n)` algorithm - which at least does not make things worse.

I do agree that doing this on the internal representation might be more efficient. Both thanks to the "support file index" optimization, and also because we don't need to wrap all non-qualifying results into `SBLineEntry`/`std::shared_ptr`s.

I decided against this, because to do so I would have to expose a new API as part of the `SB*` API layer, and afaik everything which I add there is essentially cast in stone / enshrined in the ABI, and I do expect changes as part of the "Go To" functionality.

Or is there some way how I can expose an API without immediately running into the shackles of ABI-compatibility?

Comment (vogelsgesang) [reply lldb/tools/lldb-dap/lldb-dap.cpp:1054]: > What's the largest file you've tested it on?

20 lines 🙈

File: lldb/tools/lldb-dap/lldb-dap.cpp
Comment (labath) [inline lldb/tools/lldb-dap/lldb-dap.cpp:1070]: FWIW, these strings are coming from lldb's string pool, so you should be able to just compare the raw pointers for equality.

**Review by labath — APPROVED** [review]

File: lldb/tools/lldb-dap/lldb-dap.cpp
Comment (vogelsgesang) [reply lldb/tools/lldb-dap/lldb-dap.cpp:1053]: > I would still encourage you to try this out on a large file, just to make sure it doesn't do something egregious. In llvm, SemaExpr.cpp is one of the largest files (but you can also find larger ones in the outside world).

I tried it on `SemaExpr.cpp`. Answering the `breakpointLocations` request takes something between 60 and 70 milliseconds, timed via `tail -f ~/lldb-dap.log | ts %.S`

**Comment by Michael137** [comment]
> LLVM Buildbot has detected a new failure on builder `lldb-aarch64-ubuntu` running on `linaro-lldb-aarch64-ubuntu` while building `lldb` at step 6 "test".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/59/builds/8268
> 
> Here is the relevant piece of the build log for the reference
> ```
> Step 6 (test) failure: build (failure)
> ...
> PASS: lldb-api :: python_api/target/TestTargetAPI.py (1104 of 2050)
> PASS: lldb-api :: symbol_ondemand/breakpoint_source_regex/TestSourceTextRegexBreakpoint.py (1105 of 2050)
> PASS: lldb-api :: symbol_ondemand/breakpoint_language/TestBreakpointLanguageOnDemand.py (1106 of 2050)
> PASS: lldb-api :: test_utils/TestDecorators.py (1107 of 2050)
> PASS: lldb-api :: terminal/TestSTTYBeforeAndAfter.py (1108 of 2050)
> PASS: lldb-api :: test_utils/TestInlineTest.py (1109 of 2050)
> PASS: lldb-api :: test_utils/TestPExpectTest.py (1110 of 2050)
> PASS: lldb-api :: test_utils/base/TestBaseTest.py (1111 of 2050)
> PASS: lldb-api :: symbol_ondemand/shared_library/TestSharedLibOnDemand.py (1112 of 2050)
> UNSUPPORTED: lldb-api :: tools/lldb-dap/breakpoint-events/TestDAP_breakpointEvents.py (1113 of 2050)
> FAIL: lldb-api :: tools/lldb-dap/breakpoint/TestDAP_breakpointLocations.py (1114 of 2050)
> ******************** TEST 'lldb-api :: tools/lldb-dap/breakpoint/TestDAP_breakpointLocations.py' FAILED ********************
> Script:
> --
> /usr/bin/python3.10 /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/llvm-project/lldb/test/API/dotest.py -u CXXFLAGS -u CFLAGS --env LLVM_LIBS_DIR=/home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/./lib --env LLVM_INCLUDE_DIR=/home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/include --env LLVM_TOOLS_DIR=/home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/./bin --arch aarch64 --build-dir /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/lldb-test-build.noindex --lldb-module-cache-dir /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/lldb-test-build.noindex/module-cache-lldb/lldb-api --clang-module-cache-dir /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/lldb-test-build.noindex/module-cache-clang/lldb-api --executable /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/./bin/lldb --compiler /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/./bin/clang --dsymutil /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/./bin/dsymutil --make /usr/bin/make --llvm-tools-dir /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/./bin --lldb-obj-root /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/tools/lldb --lldb-libs-dir /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/build/./lib /home/tcwg-buildbot/worker/lldb-aarch64-ubuntu/llvm-project/lldb/test/API/tools/lldb-dap/breakpoint -p TestDAP_breakpointLocations.py
> --
> Exit Code: 1
> 
> Command Output (stdout):
> --
> lldb version 20.0.0git (https://github.com/llvm/llvm-project.git revision 4f48a81a620bc9280be4780f3554cdc9bda55bd3)
>   clang revision 4f48a81a620bc9280be4780f3554cdc9bda55bd3
>   llvm revision 4f48a81a620bc9280be4780f3554cdc9bda55bd3
> Skipping the following test categories: ['libc++', 'dsym', 'gmodules', 'debugserver', 'objc']
> ========= DEBUG ADAPTER PROTOCOL LOGS =========
> 1731780632.250944853 --> 
> Content-Length: 344
> 
> {
>   "arguments": {
>     "adapterID": "lldb-native",
>     "clientID": "vscode",
>     "columnsStartAt1": true,
>     "linesStartAt1": true,
>     "locale": "en-us",
>     "pathFormat": "path",
>     "sourceInitFile": false,
>     "supportsRunInTerminalRequest": true,
>     "supportsStartDebuggingRequest": true,
>     "supportsVariablePaging": true,
>     "supportsVariableType": true
>   },
>   "command": "initialize",
>   "seq": 1,
>   "type": "request"
> }
> 1731780632.252939463 <-- 
> Content-Length: 1631
> ```

This is also failing on the macOS CI: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/15438/execution/node/97/log/
```
======================================================================
FAIL: test_column_breakpoints (TestDAP_breakpointLocations.TestDAP_setBreakpoints)
   Test retrieving the available breakpoint locations.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/tools/lldb-dap/breakpoint/TestDAP_breakpointLocations.py", line 77, in test_column_breakpoints
    self.assertEqual(
AssertionError: Lists differ: [{'co[70 chars]e': 41}, {'column': 3, 'line': 42}, {'column': 18, 'line': 42}] != [{'co[70 chars]e': 42}, {'column': 18, 'line': 42}]

First differing element 2:
{'column': 3, 'line': 41}
{'column': 3, 'line': 42}

First list contains 1 additional elements.
First extra element 4:
{'column': 18, 'line': 42}

  [{'column': 39, 'line': 40},
   {'column': 51, 'line': 40},
-  {'column': 3, 'line': 41},
   {'column': 3, 'line': 42},
   {'column': 18, 'line': 42}]
Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
----------------------------------------------------------------------
Ran 1 test in 1.554s

FAILED (failures=1)
```

Mind taking a look?

**Comment by Michael137** [comment]
Revert in https://github.com/llvm/llvm-project/commit/ceeb08b9e0a51a4d2e0804baeb579fe8a6485885 for now to unblock the bot

**Pull request merged**
